naffka.go

```diff
@@ -0,0 +1,341 @@
+		n.topics[topicName] = &topic{
+			topicName:  topicName,
+			nextOffset: offset + 1,
+		}
+	}
+	return n, nil
+}
+
+// A Message is used internally within naffka to store messages.
+// It is converted to a sarama.ConsumerMessage when exposed to the
+// public APIs to maintain API compatibility with sarama.
+type Message struct {
+	Topic     string
+	Offset    int64
+	Key       []byte
+	Value     []byte
+	Timestamp time.Time
+}
+
+func (m *Message) consumerMessage() *sarama.ConsumerMessage {
+	return &sarama.ConsumerMessage{
+		Topic:     m.Topic,
+		Offset:    m.Offset,
+		Key:       m.Key,
+		Value:     m.Value,
+		Timestamp: m.Timestamp,
+	}
+}
+
+// A Database is used to store naffka messages.
+// Messages are stored so that new consumers can access the full message history.
+type Database interface {
+	// StoreMessages stores a list of messages.
+	StoreMessages(messages []Message) error
+	// FetchMessages fetches all messages with an offset greater than startOffset
+	// and less than endOffset.
+	FetchMessages(topic string, startOffset, endOffset int64) ([]Message, error)
+	// MaxOffsets returns the maximum offset for each topic.
+	MaxOffsets() (map[string]int64, error)
+}
+
+// SendMessage implements sarama.SyncProducer
+func (n *Naffka) SendMessage(msg *sarama.ProducerMessage) (partition int32, offset int64, err error) {
+	err = n.SendMessages([]*sarama.ProducerMessage{msg})
+	return msg.Partition, msg.Offset, err
+}
+
+// SendMessages implements sarama.SyncProducer
+func (n *Naffka) SendMessages(msgs []*sarama.ProducerMessage) error {
+	byTopic := map[string][]*sarama.ProducerMessage{}
+	for _, msg := range msgs {
+		byTopic[msg.Topic] = append(byTopic[msg.Topic], msg)
+	}
+	var topicNames []string
+	for topicName := range byTopic {
+		topicNames = append(topicNames, topicName)
+	}
+
+	now := time.Now()
+	topics := n.getTopics(topicNames)
+	for topicName := range byTopic {
+		if err := topics[topicName].send(now, byTopic[topicName]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *Naffka) getTopics(topicNames []string) map[string]*topic {
+	n.topicsMutex.Lock()
+	defer n.topicsMutex.Unlock()
+	result := map[string]*topic{}
+	for _, topicName := range topicNames {
+		t := n.topics[topicName]
+		if t == nil {
+			// If the topic doesn't already exist then create it.
+			t = &topic{db: n.db, topicName: topicName}
+			n.topics[topicName] = t
+		}
+		result[topicName] = t
+	}
+	return result
+}
+
+// Topics implements sarama.Consumer
+func (n *Naffka) Topics() ([]string, error) {
+	n.topicsMutex.Lock()
+	defer n.topicsMutex.Unlock()
+	var result []string
+	for topic := range n.topics {
+		result = append(result, topic)
+	}
+	return result, nil
+}
+
+// Partitions implements sarama.Consumer
+func (n *Naffka) Partitions(topic string) ([]int32, error) {
+	// Naffka stores a single partition per topic, so this always returns a single partition ID.
+	return partitions[:], nil
+}
+
+var partitions = [1]int32{0}
```

Comment:
I've got rid of since I'd rather not bother with optimisations like this if the placement of the constant is going to be controversial.